htpasswd.go

```diff
@@ -177,30 +216,66 @@ func (h *HtpasswdFile) ReloadCheck() error {
 func (h *HtpasswdFile) Validate(user string, password string) bool {
 	_ = h.ReloadCheck()
 
+	hash := sha256.New()
+	// hash.Write can never fail
+	_, _ = hash.Write([]byte(user))
+	_, _ = hash.Write([]byte(":"))
+	_, _ = hash.Write([]byte(password))
+
 	h.mutex.Lock()
-	realPassword, exists := h.Users[user]
+	// avoid race conditions with cache replacements
+	cache := h.cache
+	realPassword, exists := h.users[user]
+	entry, cacheExists := h.cache[user]
 	h.mutex.Unlock()
 
 	if !exists {
 		return false
 	}
 
+	if cacheExists && subtle.ConstantTimeCompare(entry.verifier, hash.Sum(nil)) == 1 {
+		h.mutex.Lock()
+		// repurpose mutex to prevent concurrent cache updates
+		// extend cache entry
+		cache[user] = cacheEntry{
+			verifier: entry.verifier,
+			expiry:   time.Now().Add(PasswordCacheDuration),
+		}
+		h.mutex.Unlock()
+		return true
+	}
+
 	var shaRe = regexp.MustCompile(`^{SHA}`)
 	var bcrRe = regexp.MustCompile(`^\$2b\$|^\$2a\$|^\$2y\$`)
 
+	isValid := false
+
 	switch {
 	case shaRe.MatchString(realPassword):
 		d := sha1.New()
 		_, _ = d.Write([]byte(password))
 		if realPassword[5:] == base64.StdEncoding.EncodeToString(d.Sum(nil)) {
-			return true
+			isValid = true
 		}
 	case bcrRe.MatchString(realPassword):
 		err := bcrypt.CompareHashAndPassword([]byte(realPassword), []byte(password))
 		if err == nil {
-			return true
+			isValid = true
 		}
 	}
```

Comment:
This switch, instead of setting a true signal, should be moved into a function that returns true/false so you can just use it as needed.

Reply:
Is changed.